### PR TITLE
CDAP-9165 Add metrics to preview

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-9165

Most of the change is to extract the method from MetricsHandler to a helper class so preview can use the same logic for the metrics REST endpoint, also updated the unit test. 
Note there is a bug for the metric count in data streams pipeline, and the fix is in #8455 , the build will always fail until that pr is merged.

### CODE REVIEW

**Comment by gsps1** [comment]
in the preview endpoint we already have namespace and appId (previewId) vs Metrics API where everything has to be specified via the REST API query params.. if we are not including namespaceId and appId from pathParams into the metrics tags, we should do that.

**Comment by yaojiefeng** [comment]
Build is running: http://builds.cask.co/browse/CDAP-DUT5614-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java [hunk 1]
```diff
@@ -222,4 +298,37 @@ private ProgramRunId getProgramRunId(String namespaceId, String previewId) throw
...
+    String previewTag = APP_STRING + ":" + previewId;
+    for (String tag : tags) {
+      if (tag.startsWith(NAMESPACE_STRING)) {
+        newTags.add(namespaceTag);
+      } else if (tag.startsWith(APP_STRING)) {
+        newTags.add(previewTag);
+      } else {
+        newTags.add(tag);
+      }
+    }
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java:318]: so this handles the case where the namespace is provided in tags, but we need to add tags ourselves for namespace and app id if they are not provided as tags (as they are available in path parameter).

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java [hunk 2]
```diff
@@ -63,9 +69,12 @@
 @Singleton
 @Path(Constants.Gateway.API_VERSION_3 + "/namespaces/{namespace-id}")
 public class PreviewHttpHandler extends AbstractLogHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(PreviewHttpHandler.class);
   private static final Gson GSON = new GsonBuilder()
     .registerTypeAdapter(BasicThrowable.class, new BasicThrowableCodec()).create();
   private static final Type STRING_LIST_MAP_TYPE = new TypeToken<Map<String, List<String>>>() { }.getType();
+  private static final String NAMESPACE_STRING = "namespace";
+  private static final String APP_STRING = "app";
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java:77]: use `co.cask.cdap.common.conf.Constants.Metrics.Tag#NAMESPACE`

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java [hunk 3]
```diff
@@ -222,4 +298,37 @@ private ProgramRunId getProgramRunId(String namespaceId, String previewId) throw
...
+      }
+    }
+    return newTags;
+  }
+
+  private void overrideQueries(Map<String, MetricsQueryHelper.QueryRequestFormat> query, String namespaceId,
+                               String previewId) {
+    for (MetricsQueryHelper.QueryRequestFormat format : query.values()) {
+      Map<String, String> tags = format.getTags();
+      if (!tags.containsKey(NAMESPACE_STRING) || !tags.get(NAMESPACE_STRING).equals(namespaceId)) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java:332]: i see thats handled here for queries.we need to handle for search too

**Comment by gsps1** [comment]
@sagarkapare why don't we have a `PreviewHttpHandlerTest` ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java (hunk 2 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java:77]: That constant has value "ns", but here user will pass tags as "namespace", so I cannot use that constant.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java (hunk 1 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java:318]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java:332]: Fixed

**Comment by yaojiefeng** [comment]
closing this and reopen against release/4.1

**Pull request closed without merging**